[migration] Mapping Layer - 1. A mapping from predicates/priorities to plugin configuration

ahg-g opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Implements a registry that maps predicates/priorities to framework plugins. This PR only implements the mapping, but does not use it. A followup PR will use this mapping to produce a framework configuration that redirects mapped predicates/priorites to be executed as plugins.

**Which issue(s) this PR fixes**:
Part of #82708

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

liu-cong commented:
nit: typo pluing -> plugin

hex108 commented on pkg/scheduler/framework/plugins/default_registry.go:
```diff
@@ -31,3 +31,22 @@ func NewDefaultRegistry() framework.Registry {
+// 	if pluginSet == nil {
+// 		pluginSet = &config.PluginSet{}
+// 	}
+// 	pluginSet.Enabled = append(pluginSet.Enabled, config.Plugin{Name: name, Weight: weight})
```
It seems no need to add them to enabled plugin sets when registering predicate/priority functions, it just means that admins could use them by configuring.

ahg-g replied:
This is not for registering the plugins in the default registry, this is for registering a config producer that creates a configuration for a mapped predicate/priority.

ahg-g commented:
/assign @Huang-Wei

alculquicondor commented on pkg/scheduler/factory/plugins.go:
```diff
@@ -632,3 +640,19 @@ func ListAlgorithmProviders() string {
+
+// RegisterPredicateToFrameworkConfigProducer registers a framework config producer for a predicate.
+func RegisterPredicateToFrameworkConfigProducer(name string, producer FrameworkConfigProducer) {
+	schedulerFactoryMutex.Lock()
```
These 2 functions can be refactored to call a non-exported function that takes the map as argument.

ahg-g replied:
Done.

alculquicondor commented on pkg/scheduler/factory/plugins_test.go:
```diff
@@ -136,3 +137,104 @@ func TestBuildScoringFunctionShapeFromRequestedToCapacityRatioArgumentsNilResour
+			t.Fatalf("wanted to find %q in the configurator map.", predicate)
+		}
+		plugins, _ := producer()
+		allPlugins.Append(&plugins)
```
Shouldn't there be a public function that produces the plugins from the maps?

alculquicondor commented on pkg/scheduler/factory/plugins_test.go:
```diff
@@ -136,3 +137,104 @@ func TestBuildScoringFunctionShapeFromRequestedToCapacityRatioArgumentsNilResour
+			return
+		})
+
+	var allPlugins config.Plugins
```
gotPlugins

ahg-g replied:
Done.

alculquicondor commented on pkg/scheduler/factory/plugins_test.go:
```diff
@@ -136,3 +137,104 @@ func TestBuildScoringFunctionShapeFromRequestedToCapacityRatioArgumentsNilResour
+	var allPlugins config.Plugins
+
+	// Verify that both predicates are in the map and produce the expected filter configurations.
+	predicates := []string{testPredicateName1, testPredicateName2}
```
wantPredicates

ahg-g replied:
Done.

alculquicondor commented on pkg/scheduler/factory/plugins_test.go:
```diff
@@ -136,3 +137,104 @@ func TestBuildScoringFunctionShapeFromRequestedToCapacityRatioArgumentsNilResour
+	}
+
+	// Verify the aggregated configuration.
+	if len(allPlugins.Filter.Enabled) != len(predicates) {
```
it's simpler if you define `wantPlugins` and then use `cmp.Diff`

ahg-g replied:
Done.

alculquicondor commented on pkg/scheduler/factory/plugins_test.go:
```diff
@@ -136,3 +137,104 @@ func TestBuildScoringFunctionShapeFromRequestedToCapacityRatioArgumentsNilResour
+			len(allPlugins.Score.Enabled), len(priorities))
+	}
+	for i, score := range scorers {
+		if score != allPlugins.Score.Enabled[i].Name {
```
Is the order important?

ahg-g replied:
may not be for scores, but this is testing that append works as expected irrespective of the extension point.

alculquicondor commented on pkg/scheduler/framework/plugins/default_registry.go:
```diff
@@ -31,3 +31,22 @@ func NewDefaultRegistry() framework.Registry {
+	// priorities. For example:
+	//
+	// factory.RegisterPredicateToFrameworkConfigProducer(predicates.PodFitsHostPortsPred,
+	// 	func() (plugins config.Plugins, pluginConfig []config.PluginConfig) {
```
will we know pluginConfig at `init` time? I feel like some configurations might come from runtime sources.
In that case, you might need pluginConfig to be a parameter as well?

alculquicondor replied:
Is it just the weight?
But we could have other things for #80639

ahg-g replied:
I was mainly thinking of predicates/priorities which do not require customization, all the ones supported by DefaultProvider fall in that category. I suspect that in most (may be all) cases an empty pluginConfig will be returned, unless one wants to provide a fixed configuration. For custom predicates/priorities, we will need to explicitly register their producer, and that producer function can capture whatever parameters it wants.

In any case, we can change this from ```init``` to a function that the configurator calls explicitly with whatever arguments the custom predicates/priorities needs for their config producers.

ahg-g replied:
I will remove this for now, I think it will likely need to be in the factory package.

Huang-Wei commented on pkg/scheduler/apis/config/types.go:
```diff
@@ -209,3 +209,33 @@ type PluginConfig struct {
+
+// Append appends src Plugins to current Plugins. If a PluginSet is nil, it will
+// be created.
+func (p *Plugins) Append(src *Plugins) {
```
Does this need to be exposed to `staging` for user consuming? Or just for internal usage.

ahg-g replied:
Just internal use, kube-proxy does that as well.

Huang-Wei replied:
OK, just want to confirm on this.

ahg-g replied:
added a comment to clarify that.

Huang-Wei commented on pkg/scheduler/factory/plugins_test.go:
```diff
@@ -136,3 +137,104 @@ func TestBuildScoringFunctionShapeFromRequestedToCapacityRatioArgumentsNilResour
+func TestRegisterFrameworkConfigProducers(t *testing.T) {
+	testPredicateName1 := "testPredicate1"
+	testFilterName1 := "testFilter1"
+	RegisterPredicateToFrameworkConfigProducer(testPredicateName1,
```
As the producer map are private variables, we can't arbitrarily change their states. Instead, we should use defer() to restore to the initial state in this unit test.

ahg-g replied:
I added unregister calls in the defer function.

ahg-g replied:
now that they are private to a struct, we don't need to do that.

Huang-Wei reviewed (commented):
Some comments.

alculquicondor commented on pkg/scheduler/factory/plugins.go:
```diff
@@ -641,18 +641,46 @@ func ListAlgorithmProviders() string {
+}
+
+// ProducePredicateFrameworkConfig produces framework configuration for the list of provided
+// predicate keys. Corresponding producers should have been reigstered before calling this function.
```
s/reigstered/registered

alculquicondor commented on pkg/scheduler/factory/plugins_test.go:
```diff
@@ -179,62 +182,55 @@ func TestRegisterFrameworkConfigProducers(t *testing.T) {
+		PreFilter: &config.PluginSet{},
+		Filter: &config.PluginSet{
+			Enabled: []config.Plugin{
+				{
```
nit: inline these 3 lines

alculquicondor commented on pkg/scheduler/factory/plugins_test.go:
```diff
@@ -179,62 +182,55 @@ func TestRegisterFrameworkConfigProducers(t *testing.T) {
-				allPlugins.Score.Enabled[i].Name, score)
 
-		}
+	if !reflect.DeepEqual(wantedPlugins, gotPlugins) {
```
There's `cmp.Equal`, but you can just use the diff:

```
if diff = cmp.Diff(wantPlugins, gotPlugins); diff != "" {
  t.Errorf("unexpected plugin configuration (-want, +got): %s", diff)
}
```

ahg-g commented:
@alculquicondor @Huang-Wei I think I should place the logic in plugins.go in a standalone struct (FrameworkConfigurator or something), I will send an update shortly.

alculquicondor commented:
> Aldo Culquicondor Wei Huang I think I should place the logic in plugins.go in a standalone struct (FrameworkConfigurator or something), I will send an update shortly.

sgtm

ahg-g commented:
> > Aldo Culquicondor Wei Huang I think I should place the logic in plugins.go in a standalone struct (FrameworkConfigurator or something), I will send an update shortly.
> 
> sgtm

Done.

alculquicondor commented on pkg/scheduler/factory/framework.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+// IsPredicateRegistered returns true if the key exist in the predicates registry
+func (f *FrameworkConfigProducerRegistry) IsPredicateRegistered(name string) bool {
```
Do you expect these to be useful outside of `framework_test.go`? They can just be there or be inlined.

ahg-g replied:
yes, they will be used in the followup PR.

alculquicondor commented on pkg/scheduler/factory/framework.go:
```diff
@@ -0,0 +1,95 @@
+
+// ProducePredicateFrameworkConfig produces framework configuration for the list of provided
+// predicate keys. Corresponding producers should have been registered before calling this function.
+func (f *FrameworkConfigProducerRegistry) ProducePredicateFrameworkConfig(predicateKeys []string) (*config.Plugins, []config.PluginConfig, error) {
```
What's the benefit of having separated exported functions instead of a single `ProduceFrameworkConfig`?

ahg-g replied:
I was trying to encapsulate access to the maps so that the caller deals with the struct methods only to hide the implementation detail of the registry (this is all C++ style). In any case, I removed this because the factory will want to pass customized arguments for each producer, and so there is point in having this function anymore.

Huang-Wei commented on pkg/scheduler/factory/framework.go:
```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
2019

Huang-Wei commented on pkg/scheduler/factory/framework_test.go:
```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
2019

Huang-Wei commented:
LGTM. Thanks @ahg-g, you can squash the commits.

ahg-g commented:
> LGTM. Thanks @ahg-g, you can squash the commits.

Thanks, I made another modification: the producer now accepts a special args argument. This should make it easier to customize the produced configuration: e.g., different weights for score plugins, customize pod affinity with the CC hardPodAffinitySymmetricWeight parameter etc.

ahg-g committed:
A mapping from predicates/priorities to pluing configuration.

This PR only implements the mapping, but does not use it. A followup PR will use this mapping to produce a framework configuration that redirects mapped predicates/priorites to be exected as plugins.

Huang-Wei commented on pkg/scheduler/framework/plugins/default_registry_test.go:
```diff
@@ -0,0 +1,135 @@
+
+	args := FrameworkConfigProducerArgs{Weight: 1}
+	predicatePlugins, _, err := produceFrameworkConfig(
+		([]string{testPredicateName1, testPredicateName2}), registry.PredicateToFrameworkConfigProducer, args)
```
```suggestion
		[]string{testPredicateName1, testPredicateName2}, registry.PredicateToFrameworkConfigProducer, args)
```

Huang-Wei commented on pkg/scheduler/framework/plugins/default_registry.go:
```diff
@@ -31,3 +34,45 @@ func NewDefaultRegistry() framework.Registry {
 	}
 }
+
+// FrameworkConfigProducerArgs contains arguments that are passed to the producer.
```
It'd be good to comment that we may introduce more fields in the future. Right now, only `Weight` field.

ahg-g replied:
done

Huang-Wei commented:
/lgtm

Thanks @ahg-g for bringing this up!
